types/opt/value.go

```diff
@@ -0,0 +1,167 @@
+// Copyright (c) Tailscale Inc & AUTHORS
+// SPDX-License-Identifier: BSD-3-Clause
+
+package opt
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"reflect"
+)
+
+// Value is an optional value to be JSON-encoded.
+// Unlike the specialized Bool type, an empty (unset) Value[T] is marshaled as
+// a JSON null regardless of the "omitempty" JSON tag option when using `encoding/json`.
+// With https://github.com/go-json-experiment/json, however, empty optional Values are
+// omitted from the output.
```

Comment:
Is this saying you intend for these types to JSON marshal different ways depending on what JSON encoder is used?

Add some tests for both JSON encoders?

But that seems like maybe not a desirable property, as we're trying to move to the json-experiment encoder in more places but keep the same encoding.

cc @dsnet

Reply:
It's not that I intend for these types to marshal differently depending on the JSON encoder used, but rather that the new json-experiment encoder has a different [default behavior](https://github.com/go-json-experiment/json?tab=readme-ov-file#behavior-changes) when the `omitempty` JSON tag option is used:

> When marshaling a Go struct, a struct field marked as omitempty is omitted if the field value would encode as an empty JSON value, which is defined as a JSON null, or an empty JSON string, object, or array.

Reply:
I added a few more tests. The `null_for_unset_jsonv2` produces the same output as `null_for_unset`, but `null_for_unset_omitempty` and `null_for_unset_omitempty_jsonv2` result in different JSON outputs as `json-experiment` omits empty (unset) `Value[T]`s.